### PR TITLE
Refactor Adapters

### PR DESCRIPTION
**Pull request by marcelklehr**
@jlbprof I've decided to not implement a change list after all, as I need to keep track of the ids of the created items. For that it'd be better to have return values for each operations, ergo I used methods again for every operation. However, I realize that the caching you did in the WebDAV adapter could potentially be useful to other adapters, so I created an abstract CachingAdapter that already implements all the operations that work with the cache and only leaves the population and persistence of the cache to you.

I've also generalized the adapter operations into a Resource interface which the interface for the local tree inherits from as well, just for consistency. So, the sync step could potentially sync any two implementations of Resource.

### CODE REVIEW

**Comment by jlbprof** [comment]
I will look at it in the morning.  From your description this sounds reasonable.

File: src/lib/adapters/CachingAdapter.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
+import Tree from '../Tree'
+import Adapter from '../Adapter'
+
+export default class CachingAdapter {
```
Comment (jlbprof) [inline src/lib/adapters/CachingAdapter.js]: Did you want CachingAdapter to inherit from Adapter?

File: src/lib/adapters/CachingAdapter.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  constructor(server) {
+    this.highestId = 0
+  }
+
+  async getBookmarksTree() {
+    return this.bookmarksCache
```
Comment (jlbprof) [inline src/lib/adapters/CachingAdapter.js]: I do not see how bookmarksCache is getting set.

File: src/lib/LocalTree.js [hunk 3]
```diff
@@ -0,0 +1,265 @@
+import browser from './browser-api'
+import Bookmark from './Bookmark'
```
Comment (jlbprof) [inline src/lib/LocalTree.js]: Bookmark no longer exists in the src tree.  Do an import of Tree to get Bookmark.

File: src/lib/Tree.js [hunk 4]
```diff
@@ -1,265 +1,71 @@
...
-
-  async createNode(bookmark) {
-    console.log('CREATE', bookmark)
-
-    if (this.getBookmarkById(bookmark.id)) {
-      throw new Error(
-        'trying to create a node for a bookmark that already has one'
+import Crypto from './Crypto'
+
+export default class Bookmark {
```
Comment (jlbprof) [inline src/lib/Tree.js]: Bookmark has been moved to the Tree.js file

**Comment by marcelklehr** [comment]
Oh, yeah. It's definitely not finished :D I should have mentioned that :sweat_smile:

File: src/lib/adapters/CachingAdapter.js (hunk 2 above)
Comment (marcelklehr) [reply src/lib/adapters/CachingAdapter.js]: The adapter that inherits from the caching adapter implements onSyncStart and onSyncComplete in order to set the cache. I could also add a method for setting the cache... *shrug*

**Comment by marcelklehr** [comment]
I'd say it's ready for more extensive testing now. I haven't rewritten the WebDAV adapter, yet, but that should be trivial with the CachingAdapter as a base. NextcloudAdapter and FakeAdapter (for tests) are both in a working state.

**Pull request merged**